KeyDistribution_BB84/ReferenceImplementation.qs

```diff
@@ -75,111 +76,111 @@ namespace Quantum.Kata.KeyDistribution {
 
-        return count / IntAsDouble(Length(keyA)) >= p;
+        return IntAsDouble(count) / IntAsDouble(N) >= IntAsDouble(threshold) / 100.0;
     }
 
 
-    operation Task26_BB84_Reference(qs : Qubit[],  p : Double) : Bool[] {
-        // 1. Choose random basis and bits to encode
-		let basis = RandomArray_Reference(Length(qs));
-		let bits = RandomArray_Reference(Length(qs));
-		
-		// 2. Alice prepares her qubits
-		PrepareAlicesQubits_Reference(qs, basis, bits);
+    // Task 2.6. Putting it all together 
+    operation T26_BB84Protocol_Reference () : Unit {
+        let threshold = 99;
+
+        using (qs = Qubit[20]) {
+            // 1. Choose random basis and bits to encode
+            let basesAlice = RandomArray_Reference(Length(qs));
+            let bitsAlice = RandomArray_Reference(Length(qs));
         
-		// 3. Bob chooses random basis to measure in
-        let bBob = RandomArray_Reference(Length(qs));
+            // 2. Alice prepares her qubits
+            PrepareAlicesQubits_Reference(qs, basesAlice, bitsAlice);
+        
+            // 3. Bob chooses random basis to measure in
+            let basesBob = RandomArray_Reference(Length(qs));
+
+            // 4. Bob measures Alice's qubits'
+            let bitsBob = MeasureBobsQubits_Reference(qs, basesBob);
 
-		// 4. Bob measures Alice's qubits'
-        let mBob = MeasureBobsQubits_Reference(qs, bBob);
+            // 5. Generate shared key
+            let keyAlice = GenerateSharedKey_Reference(basesAlice, basesBob, bitsAlice);
+            let keyBob = GenerateSharedKey_Reference(basesAlice, basesBob, bitsBob);
 
-		// 5. Generate shared key
-        let keyA = Task24_GenerateKey_Reference(basis, bBob, bits);
-        let keyB = Task24_GenerateKey_Reference(basis, bBob, mBob);
+            // 6. Ensure at least the minimum percentage of bits match
+            if (CheckKeysMatch_Reference(keyAlice, keyBob, threshold)) {
+                Message($"Successfully generated keys {keyAlice}/{keyBob}");
+            }
 
-		// 6. Ensure at least the minimum percentage of bits match
-        if (Task25_CheckKeysMatch_Reference(keyA, keyB, p)) {
-            return keyA;
+            ResetAll(qs);
         }
-        return new Bool[0];
     }
 
 
     //////////////////////////////////////////////////////////////////
     // Part III. Eavesdropping
     //////////////////////////////////////////////////////////////////
 
-    operation Task31_Reference (q : Qubit, basis : Bool) : Int {
-        mutable res = 0;
-        if (not basis) {
-            // Measure in the rectangular basis
-            set res = M(q) == Zero ? 0 | 1;
-        } else {
-            // Measure in the diagonal basis
-            H(q);
-            set res = M(q) == Zero ? 0 | 1;
-            H(q);
-        }
-        return res;
+    // Task 3.1. Eavesdrop!
+    operation Eavesdrop_Reference (q : Qubit, basis : Bool) : Bool {
+        return ResultAsBool(Measure([basis ? PauliX | PauliZ], [q]));
```

Comment:
Hi @tcNickolas 
I just worked through this kata and have trouble understanding why in the reference implementation, the measurement is being done as `basis ? PauliX | PauliZ` instead of `basis ? PauliZ | PauliX` as should be if the definition of `basis` is consistent throughout this kata?

I am leaving another comment where I had an issue with this.

Reply:
I think this is consistent with the use of basis throughout the kata? `basis = false` indicates |0⟩/|1⟩ basis, and to do a measurement in it you need to measure in Pauli Z basis (since these states are eigenstates of Pauli Z matrix); `basis = true` indicates |+⟩/|-⟩ basis, which corresponds to measurement in Pauli X basis.

Reply:
Thanks, I think I was confused about which of Pauli Z or X maps to the said basis states. Now, it's clear. :)